bot/utils/redis_dict.py

```diff
@@ -0,0 +1,136 @@
+from __future__ import annotations
+
+import json
+from collections.abc import MutableMapping
+from enum import Enum
+from typing import Dict, List, Optional, Tuple, Union
+
+import redis as redis_py
+
+from bot import constants
+
+ValidRedisKey = Union[str, int, float]
+JSONSerializableType = Optional[Union[str, float, bool, Dict, List, Tuple, Enum]]
+
+
+class RedisDict(MutableMapping):
+    """
+    A dictionary interface for a Redis database.
+
+    Objects created by this class should mostly behave like a normal dictionary,
+    but will store all the data in our Redis database for persistence between restarts.
+
+    Redis is limited to simple types, so to allow you to store collections like lists
+    and dictionaries, we JSON deserialize every value. That means that it will not be possible
+    to store complex objects, only stuff like strings, numbers, and collections of strings and numbers.
+    """
+
+    _namespaces = []
+    _redis = redis_py.Redis(
+        host=constants.Bot.redis_host,
+        port=constants.Bot.redis_port,
+    )  # Can be overridden for testing
+
+    def __init__(self, namespace: Optional[str] = None) -> None:
+        """Initialize the RedisDict with the right namespace."""
+        super().__init__()
+        self._has_custom_namespace = namespace is not None
+
+        if self._has_custom_namespace:
+            self._set_namespace(namespace)
+        else:
+            self.namespace = "global"
+
+    def _set_namespace(self, namespace: str) -> None:
+        """Try to set the namespace, but do not permit collisions."""
+        while namespace in self._namespaces:
+            namespace = namespace + "_"
+
+        self._namespaces.append(namespace)
+        self._namespace = namespace
```

Comment:
I think I'd prefer if there was a `clean_namespace` function or similar, that would return the namespace, and have the attr always set in the `__init__`. This way, the attr is being set in two methods (actually isn't but I think that's a bug, i.e. `namespace` vs `_namespace`), and isn't annotated anywhere.

In case of collision, the determined namespace depends on order, right? The order in which the `RedisDict` are created. Could this lead to potential problems, if a cog is renamed or replaced? I suppose the worst that can happen is the persisted data is lost. I find it a little difficult to reason about and probably just need to spend more time thinking about what we're doing here.

Reply:
I think this is a fair point, I don't mind making this change.

> Could this lead to potential problems, if a cog is renamed or moved?

It's very unlikely, but this is why _we should never use Redis to store anything important_, only for caching data we can afford to lose. With that principle firmly in place, I don't really see the need to dive too deep into those kinds of edge cases.

Reply:
With the rewrite, this feels less relevant. `_namespace` is now really only ever set in this method, and is only initialized as None in `__init__`. This is done because `__init__` happens too early in the lifecycle for us to be able to set the attribute yet. We need to wait until `__set_name__`.